sui_core/src/authority/authority_checkpoints.rs

```diff
@@ -0,0 +1,1229 @@
+                .filter_map(|((seq, tx), in_chk)| {
+                    if in_chk.is_none() {
+                        Some((tx, seq))
+                    } else {
+                        None
+                    }
+                }),
+        )?;
+
+        // Write to the database.
+        batch.write()?;
+
+        Ok(self.lowest_unprocessed_checkpoint())
+    }
+}
+
+pub struct FragmentReconstruction {
+    pub committee: Committee,
+    pub global: GlobalCheckpoint<AuthorityName, TransactionDigest>,
+}
+
+impl FragmentReconstruction {
+    /// Take an ordered list of fragments and attempts to construct a connected
+    /// component checkpoint with weight over 2/3 of stake. Note that the minimum
+    /// prefix of links is used in this process.
+    pub fn construct(
+        seq: u64,
+        committee: Committee,
+        fragments: &[CheckpointFragment],
+    ) -> Result<Option<FragmentReconstruction>, SuiError> {
+        let mut span = SpanGraph::new(&committee);
+        let mut fragments_used = Vec::new();
+        let mut proposals: HashMap<AuthorityName, CheckpointSummary> = HashMap::new();
+
+        for frag in fragments {
+            // Double check we have only been given waypoints for the correct sequence number
+            debug_assert!(frag.diff.first.waypoint.sequence_number == seq);
+
+            // Check the checkpoint summary of the proposal is the same as the previous one.
+            // Otherwise ignore the link.
+            let n1 = &frag.proposer.0.authority;
+            if *proposals
+                .entry(*n1)
+                .or_insert_with(|| frag.proposer.0.checkpoint.clone())
+                != frag.proposer.0.checkpoint
+            {
+                continue;
+            }
+
+            let n2 = &frag.other.0.authority;
+            if *proposals
+                .entry(*n2)
+                .or_insert_with(|| frag.other.0.checkpoint.clone())
+                != frag.other.0.checkpoint
+            {
+                continue;
+            }
+
+            // Add to the links we will consider.
+            fragments_used.push(frag);
+
+            // Merge the link.
+            let (top, weight) = span.merge(n1, n2);
+
+            // We have found a connected component larger than the 2/3 threshold
+            if weight >= committee.quorum_threshold() {
+                // Get all links that are part of this component
+                let mut active_links: VecDeque<_> = fragments_used
+                    .into_iter()
+                    .filter(|frag| span.top_node(&frag.proposer.0.authority).0 == top)
+                    .collect();
+
+                let mut global = GlobalCheckpoint::new(seq);
+                while let Some(link) = active_links.pop_front() {
+                    match global.insert(link.diff.clone()) {
+                        Ok(_) | Err(WaypointError::NothingToDo) => {} // Do nothing
+                        Err(WaypointError::CannotConnect) => {
```

Comment:
So here is a common scenario in which this can happen (in which for simplicity all validators are honest):
- Each validator out of 2f+1 makes a proposal.
- Strictly enough fragments are submitted to make a checkpoint with 2f+1 proposals. No fragment connects the remaining proposals to the checkpoint.
- However, this means that f validators cannot connect their proposal to the resulting checkpoint, leading to the error above.

1) Yet, we are sure that f+1 honest validators can connect and can therefore lead to the generation of a good collectivelly signed checkpoint. 2) I have not added in `attempt_to_construct_checkpoint` a second strategy that looks at any fragments that the node has into the checkpoint (that are not part of the checkpoint) to try to reconstruct it.